machineconfiguration/v1alpha1/types_machineosbuild.go

```diff
@@ -0,0 +1,174 @@
+type PriorMachineOSBuilds struct {
+	// name is the name of the build
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+	// buildFailure contains an optional message of why this build ended prematurely.
+	// +optional
+	BuildFailure string `json:"buildFailure"`
+}
+
+// Refers to the name of a MachineConfigPool (e.g., "worker", "infra", etc.):
+type MachineConfigPoolReference struct {
+	// name is the name of the referenced object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of a rendered MachineConfig (e.g., "rendered-worker-ec40d2965ff81bce7cd7a7e82a680739", etc.):
+type RenderedMachineConfigReference struct {
+	// name is the name of the referenced object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of a (future) MachineOSImage (e.g., "worker-os-image-167651b10ec98af17971d6a47df9e22f", etc.):
+type MachineOSImageReference struct {
+	// name is the name of the referenced object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of an image registry push/pull secret needed in the build process.
+type ImageSecretObjectReference struct {
+	// name is the name of the referenced object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// BuildInputs holds all of the information needed to trigger a build
+type BuildInputs struct {
+	// imageBuilderType specifies the backend to be used to build the image.
+	// Valid options are: OpenShiftImageBuilder, PodImageBuilder, and Default (OpenShiftImageBuilder)
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
+	// baseOSImageURL is the base OSImage we use to build our custom image.
+	// +kubebuilder:validation:Required
+	BaseOSImageURL string `json:"baseOSImageURL"`
+	// baseImagePullSecret is the secret used to pull the base image.
+	// +kubebuilder :validation:Required
+	BaseImagePullSecret string `json:"baseImagePullSecret"`
+	// finalImagePushSecret is the secret used to connect to a user registry.
+	// +kubebuilder:validation:Required
+	FinalImagePushSecret ImageSecretObjectReference `json:"finalImagePushSecret"`
+	// finalImagePullSecret is the secret used to pull the final produced image.
+	// +kubebuilder:validation:Required
+	FinalImagePullSecret ImageSecretObjectReference `json:"finalImagePullSecret"`
+	// containerFile describes the custom data the user has specified to build into the image.
+	// +optional
+	Containerfile []byte `json:"containerFile"`
+	// finalImagePullspec describes the location of the final image.
+	// +kubebuilder:validation:Pattern=`^https://`
+	// +kubebuilder:validation:Required
+	FinalImagePullspec string `json:"finalImagePullspec"`
+}
+
+type MachineOSImageBuilderType string
+
+const (
+	OCPBuilder     MachineOSImageBuilderType = "OpenShiftImageBuilder"
+	PodBuilder     MachineOSImageBuilderType = "PodImageBuilder"
+	DefaultBuilder MachineOSImageBuilderType = "Default"
```

Comment:
question: I presume we'll make the link between the `DefaultBuilder` and `OCPBuilder` within the MCO?

Reply:
yep!